File: substring.py
#
# https://www.geeksforgeeks.org/length-of-the-longest-substring-without-repeating-characters/
#
# Python3 program to find the length
# of the longest substring
# without repeating characters
def longestUniqueSubsttr(string):
    # last index of every character
    last_idx = {}
    max_len = 0
    
    # starting index of current
    # window to calculate max_len
    start_idx = 0
    
    for i in range(0, len(string)):
        print("checking", string[i])
		# Find the last index of str[i]
		# Update start_idx (starting index of current window)
		# as maximum of current value of start_idx and last
		# index plus 1
        if string[i] in last_idx:
           print(string[i],"repeat")
           start_idx = max(start_idx, last_idx[string[i]] + 1)
        
        print("start_idx", start_idx)
		    # Update result if we get a larger window
        max_len = max(max_len, i-start_idx + 1)

		    # Update last index of current char.
        last_idx[string[i]] = i
        print(last_idx)
    return max_len


# Driver program to test the above function
string = "geeksforgeeks"
print("The input string is " + string)
length = longestUniqueSubsttr(string)
print("The length of the longest non-repeating character" +
	" substring is " + str(length))

File: substring.c
//
// based on https://www.geeksforgeeks.org/length-of-the-longest-substring-without-repeating-characters/
//
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#define MAX(a,b)            (((a) > (b)) ? (a) : (b))

int substrings(char *string) {
    int i,j;
    int len = strlen(string);
    int res = 0;
    int visited[256];
    
    for (i = 0; i < len; i++) {
        for (j = i; j < len; j++) {
            printf("%d %d checking %c \n ", i,j,string[j]);
            if (visited[string[j]] == 1) {
                printf("visited\n");
                printf(" break\n");
                break;
            } else {
                res = MAX(res, j - i + 1);
                visited[string[j]] = 1;
                printf("unique, set visited %c to 1, length: %d\n", string[j], res);
            }
        }
        visited[string[i]] = 0;
        printf("clearing %c\n", string[i]);
    }
    return res;
}

int main() {
    char input[] = "geeksforgeeks";
    
    printf("%d", substrings(input));
    return 0;
}
